Type script

aminya opened the pull request:
# Ready to get merged -  Tested and works as expected

- [X] Converted the codebase (JS/CS) to TypeScript. 
- [X] Changed atom-space-pen-views to atom-select-list
- [X] Defining Selector Class instead of module
- [X] using a StatusBar Class instead of module
Fixes https://github.com/JunoLab/atom-indent-detective/issues/15 and Fixes https://github.com/JunoLab/atom-indent-detective/issues/16
- [X] Optimized the code. 
  - various optimizations
  - Optimized for loops using (see [TypeScript-Optimization](https://github.com/aminya/TypeScript-optimization)):
	  - Not looking up arrays in for-head
      - Using ts-ransform for-of plugin
- [X] build using Webpack

- [X] Loading time is cut from **71ms to 0ms!**

before:

![image](https://user-images.githubusercontent.com/16418197/76260587-23601480-6226-11ea-83a3-36ac98c4e020.png)

after:

![image](https://user-images.githubusercontent.com/16418197/78467529-630a0700-76d3-11ea-8820-6a09b6e46d7f.png)

- [X] Added configuration for TypeScript
- [X] Added type declaration/annotations for some of the variables
- [X] Added TODOs for improving the code
- [x] Added Tests
- [x] Added Benchmarks

`/src` includes the typescript source files (input).
`/lib` includes the compiled javascript (output). 


- Squashed and remade the history based on @aviatesk suggestion.

pfitzseb commented:
Neat, let me know when you think this is done.

Aside: I'd very much prefer no semi-colons -- they're not very useful in JS, and TS has even [fewer corner cases where `;`s do something](https://medium.com/@eugenkiss/dont-use-semicolons-in-typescript-474ccfe4bdb3).

aminya commented:
> 
> 
> Neat, let me know when you think this is done.

I am facing a problem with importing a class from `atom-select-list`:
https://stackoverflow.com/questions/60599391/calling-javascript-class-from-typescript-uncaught-typeerror-is-not-a-const

Do you know what is going on here?

aminya commented:
> 
> 
> > Neat, let me know when you think this is done.
> 
> I am facing a problem with importing a class from `atom-select-list`:
> [stackoverflow.com/questions/60599391/calling-javascript-class-from-typescript-uncaught-typeerror-is-not-a-const](https://stackoverflow.com/questions/60599391/calling-javascript-class-from-typescript-uncaught-typeerror-is-not-a-const)
> 
> Do you know what is going on here?

I think this is fixed in https://github.com/JunoLab/atom-indent-detective/pull/17/commits/1b597d14a9494bda4a508dfec389878267ec4513, but I get a new error:
```typescript
C:\Users\yahyaaba\Documents\GitHub\atom-indent-detective\node_modules\etch\lib\dom.js:25 Uncaught Error: Invalid child node: undefined
    at dom (C:\Users\yahyaaba\Documents\GitHub\atom-indent-detective\node_modules\etch\lib\dom.js:25:15)
    at Function.dom.(anonymous function).children [as ol] (C:\Users\yahyaaba\Documents\GitHub\atom-indent-detective\node_modules\etch\lib\dom.js:64:12)
    at SelectListView.renderItems (C:\Users\yahyaaba\Documents\GitHub\atom-indent-detective\node_modules\atom-select-list\src\select-list-view.js:225:16)
    at SelectListView.render (C:\Users\yahyaaba\Documents\GitHub\atom-indent-detective\node_modules\atom-select-list\src\select-list-view.js:199:12)
    at Object.initialize (C:\Users\yahyaaba\Documents\GitHub\atom-indent-detective\node_modules\etch\lib\component-helpers.js:32:31)
    at new SelectListView (C:\Users\yahyaaba\Documents\GitHub\atom-indent-detective\node_modules\atom-select-list\src\select-list-view.js:25:10)
    at Object.selector_show (C:\Users\yahyaaba\Documents\GitHub\atom-indent-detective\dist\selector.js:11:26)
    at HTMLElement.indentDetectiveChooseIndent (../lib/indent-detective.ts:1:12)
    at CommandRegistry.handleCommandEvent (C:\Users\yahyaaba\AppData\Local\atom-nightly\app-1.46.0-nightly8\resources\app\static\<embedded>:11:349292)
    at CommandRegistry.dispatch (C:\Users\yahyaaba\AppData\Local\atom-nightly\app-1.46.0-nightly8\resources\app\static\<embedded>:11:347767)
    at HTMLSpanElement.view.onclick (C:\Users\yahyaaba\Documents\GitHub\atom-indent-detective\dist\status.js:24:34)
```

aminya commented:
> 
> 
> > > Neat, let me know when you think this is done.
> > 
> > 
> > I am facing a problem with importing a class from `atom-select-list`:
> > [stackoverflow.com/questions/60599391/calling-javascript-class-from-typescript-uncaught-typeerror-is-not-a-const](https://stackoverflow.com/questions/60599391/calling-javascript-class-from-typescript-uncaught-typeerror-is-not-a-const)
> > Do you know what is going on here?
> 
> I think this is fixed in [1b597d1](https://github.com/JunoLab/atom-indent-detective/commit/1b597d14a9494bda4a508dfec389878267ec4513), but I get a new error:
> Error:

This is solved in https://github.com/JunoLab/atom-indent-detective/pull/17/commits/bf753d49d9dab3311807c7bf891ee1bc23628d21 🚀

aminya commented:
@pfitzseb This is ready to get merged for now. 😃 

I have added some TODOs for future development.

aminya commented:
This is very ready to get merged 😄 

I fixed https://github.com/JunoLab/atom-indent-detective/issues/15 by using a StatusBar Class instead of a module.

aviatesk commented:
My thoughts on this:

- the changes are too big to be safely merged; in general it's better to split changes into small PRs as possible https://rangle.io/blog/the-art-of-humanizing-pull-requests-prs/
  * it's great that fixing #15 but it's better to be done in a separate PR
- imho TypeScript would be not so fruitful but bothersome for this package
  * we need to sync the version of TS
  * this is such a small package that we can easily track its implementation and we don't really need to rely on type information; code would be verbose
  * TS should be introduced when really necessary, and in this case it's not imho
  * (well, I'm not just object to static-typing -- we use [flow](https://flow.org/) for [Hydrogen](https://github.com/nteract/hydrogen) for example, and I feel the type checking for a package in that scale is useful; my idea is about the balance between the verboseness/bothersome vs. usefulness of it)

aminya commented:
> 
> 
> My thoughts on this:
> 
>     * the changes are too big to be safely merged; in general it's better to split changes into small PRs as possible [rangle.io/blog/the-art-of-humanizing-pull-requests-prs](https://rangle.io/blog/the-art-of-humanizing-pull-requests-prs/)
>       
>       * it's great that fixing #15 but it's better to be done in a separate PR

Thank you for your opinion.

Regarding your first point, I totally get your point regarding small PRs and I agree with it, however, this PR meant to fix the performance of this package, and for that, I had to rewrite different sections of the code, and edits in small chunks were either not possible or didn't help because of code overhaul (e.g. space-pen-views -> atom-select-list).

> 
>     * imho TypeScript would be not so fruitful but bothersome for this package
>       
>       * we need to sync the version of TS
>       * this is such a small package that we can easily track its implementation and we don't really need to rely on type information; code would be verbose
>       * TS should be introduced when really necessary, and in this case it's not imho
>       * (well, I'm not just object to static-typing -- we use [flow](https://flow.org/) for [Hydrogen](https://github.com/nteract/hydrogen) for example, and I feel the type checking for a package in that scale is useful; my idea is about the balance between the verboseness/bothersome vs. usefulness of it)


I want to disagree with this one though. 
- You don't need to do anything regarding TS. This package is good to go and will act like a JS package for usage (JS is already built).

- Typescript isn't about just tracking the implementation. It brings many improvements over simple JavaScript (enhances JS without drawbacks):
	- it reveals many things that are not visible when you use JS. Things like type-changing of the variables, if a function returns the type we want, performance improvement by using better types, etc. You see this package loads 1/4 of the original, and its runtime was better in my measurements (30-50%).
	- IntelliSense/linting
	- API documentation on the fly. It allows others to understand what the code is doing.
	- You can write your normal JS code in TypeScript and it works. TypeScript infers many of the types automatically. 
	- We can compile TypeScript to Wasm using AssemblyScript which opens a whole new world for improving the package. For example, I plan to compile the `getIndet` function of this package.
	- The future isn't JS. For example, VSCode is already using TS (which IMO has a better performance while I like atom more).

Read this article to know more about the advantages:
https://dzone.com/articles/what-is-typescript-and-why-use-it

aminya commented:
- I furthered improved the performance in https://github.com/JunoLab/atom-indent-detective/pull/17/commits/bd320dd2131887f656d970c8e9ed12a3a9641bc7 by defining a Selector class instead of a module and separating different independent functions from each other.

- Optimized for loops using (see [TypeScript-Optimization](https://github.com/aminya/TypeScript-optimization)):
	- Not looking up arrays in for-head
    - Using babel transform for-of plugin

- Minify using Babel

- Squashed and remade the history based on @aviatesk suggestion.

AzamatB commented:
Bump. Can this get merged?

pfitzseb commented on lib_src/status.ts:
```diff
@@ -0,0 +1,76 @@
+    updateDisplay(editor?: TextEditor | undefined) {
+        if (editor) {
+            this.view.style.display = ""
+            return this.updateText(editor)
```
I don't think any of this and the following methods should return anything.

aminya replied:
Fixed in https://github.com/JunoLab/atom-indent-detective/pull/17/commits/7e986ae627b9ba61e0b5ef4d48a5e95d4f12b02f

pfitzseb commented on lib_src/indent-detective.ts:
```diff
@@ -0,0 +1,264 @@
+    // if (enableDebug) {
+    //     console.log(`Indent Detective report for ${editor.getPath()}`)
+    //     console.log(counts)
+    // }
```
Let's keep this code.

pfitzseb commented on lib_src/indent-detective.ts:
```diff
@@ -0,0 +1,264 @@
+function setSettings(editor: TextEditor, length: lengthSetting) {
+    // if (enableDebug) {
+    //     console.log(`-> decided for ${length}`)
+    // }
```
Let's keep this code.

pfitzseb commented on lib_src/indent-detective.ts:
```diff
@@ -0,0 +1,264 @@
+        description:
+            "Write possible indentations that package should consider (changing requires Atom's restart/reload)",
+        order: 1
+    }
```
Let's keep the debug logging setting.

aminya replied:
Fixed in https://github.com/JunoLab/atom-indent-detective/pull/17/commits/fead62f47eccaa6c563e8f49f85259f7b750eb39

aminya commented:
> 
> 
> Alright, this looks mostly good to me. It would be great if you could change two things:
> 
>     * Semicolons aren't particularly useful, _especially_ in TS, so I'd heavily prefer not having them in the codebase.
> 

Thanks to Prettier no semicolons are used.
https://github.com/aminya/atom-indent-detective/blob/e15d983a882d1dc720f50a0fd799b82910aa8372/package.json#L47

>     * I don't think checking in the transpiled code is a good idea -- you're transpiling in `npm build`, that's enough. Would probably make sense to put all TS sources in `src` and have the JS in `lib` then.

What do you mean by checking in the code here?

travigd commented:
Drive-by comments:
* Yes, you should not be checking in transpiled code.
* The TypeScript code should live under the `src/` directory
* Remove commented-out directives from tsconfig.json, and the tsconfig.json should live in the project root, not under `src/`
* Prettier should only be run on the source code, not the generated output
* Typically, the `tsc` command is labeled `build` (so you can do `npm run build`)

Beyond that, this PR is probably too large. There are a number of things to do about this (these are options, not necessarily what **I** would do):
* Introduce prettier in a second commit
* Instead of re-writing code to TypeScript, write declaration files (`.d.ts`) that live alongside the JS and then gradually migrate to just having the code in TS (downside: requires the .d.ts file to be consistent with the .js file)
* TS can work in a JS project and parse JS files. You won't get as-strict type checking in the JS files, and sometimes the inferred types are overly broad (there will be a lot of `any` types), but it's a much more gradual migration path.

The third option definitely seems appealing to me if you're concerned about PR size. It could be:
1. Add TypeScript in `allowJS` mode, don't change anything else
2. One _file_ at a time, convert from JS to TS. This should be strictly just changing JS to TS, it shouldn't introduce new helper functions, etc.

aminya commented:
>     * Yes, you should not be checking in transpiled code.
> 

Atom (apm) packages are not built automatically like how npm packages are. We need to include them in the repository.


>     Instead of re-writing code to TypeScript, write declaration files (.d.ts) that live alongside the JS and then gradually migrate to just having the code in TS (downside: requires the .d.ts file to be consistent with the .js file)

This is an atom package, not an npm package, and so this is not the dependency of any other packages. So no one needs those `.d.ts` files. The main point of this conversion was to use the power of TypeScript, not just to introduce declaration files for others to use.


> The third option definitely seems appealing to me if you're concerned about PR size. It could be:
> 
>     Add TypeScript in allowJS mode, don't change anything else
>     One file at a time, convert from JS to TS. This should be strictly just changing JS to TS, it shouldn't introduce new helper functions, etc.
> 

That is what I am doing for https://github.com/JunoLab/atom-julia-client.

travigd commented:
> Atom (apm) packages are not built automatically like how npm packages are. We need to include them in the repository.

Quick clarification: NPM packages aren't "built automatically", but rather, they're published with the TypeScript code already having been built (so the actual thing you get when you `npm install` is just the transpiled code, no TypeScript source usually).

It seems that [Hydrogen](https://github.com/nteract/hydrogen) is able to write their code using Flow and they don't check in generated code. I'm not an expert here, but apparently apm is just a wrapper around npm, so I think the same process for shipping generated code still applies.

aviatesk commented:
fwiw, Hydrogen does _not_ any generate code. Flow is a static type checker for (kinda) plain JS.

And yes, `apm` assumes the code that will be loaded is there inside the repository unless we rely on the additional transpiler package, which should add unnecessary burden for package loading time.

aminya commented:
There is a compiler: https://github.com/smhxx/atom-ts-transpiler. But we don't wanna rely on the user's computer for compiling anyways. It makes sense to deliver the ready JS files to the user instead.


Again here like I mentioned in the Slack group, Github does a very bad job in showing the code differences. 😞 Use an external code diff like Beyond Compare that shows the differences based on the folders.

Arguing about git difference for executables (generated JS here) doesn't help much. You should either ignore reviewing it or use an external code diff.

aviatesk commented:
I meant that transpiler, of course.

> Again here like I mentioned in the Slack group, Github does a very bad job in showing the code differences. 

Maybe you just need .gitattribute then, no ?

travigd commented:
> fwiw, Hydrogen does not any generate code. Flow is a static type checker for (kinda) plain JS.

This is not accurate. Flow JS code is not runnable with a JS interpreter. For example, this code will raise many syntax errors if run in a Node interpreter. Flow code has to be transpiled (just like TS) before it can be run by a JS VM.

https://github.com/nteract/hydrogen/blob/master/lib/main.js, for example, has syntax that is not compatible with JS (notable, type annotations on function parameters and `import type ...` syntax). I'm not sure how Hydrogen actually distributes its code. I think it transpiles code on-the-fly, but not sure how that's setup.

See https://github.com/smhxx/atom-ts-transpiler for something similar for TypeScript. In the documentation, it also does assert that you only pay for compilation on installation.

> Maybe you just need .gitattribute then, no ?

GitHub has very dodgy support for `.gitattributes` (I have spent many hours trying to fight about this). GitLab is much better in this respect.

aviatesk commented:
Here is my summary of pros and cons of introducing TS into an Atom package.

Pros:
- (gradual) type checking: we can be more safe as far as we want
- extensive development features like completion, linter, etc
  * would be useful  especially for large code base like julia-client
- when one day Juno itself moves to VSCode, it _might_ help the migration if our code base is written in TS, but who knows VSCode's future API

Cons:
- boilerplates: type declarations themselves are sometimes super annoying, e.g. when you develop React code within TS, at least to some people
- incomplete (gradual) type safety: more safety and utilities we want, more boilerplates we need to accept
- additional development complexity
  * since Atom doesn't come with native TS transpiler, we need to build each time we modify our code base
  * there might be a confusion arising around the difference between acually developed code (TS) and acually loaded code (JS)
- if we only want linter, completion etc, we can just use eslint and/or throw your JS code to TS server

To me, honestly, I would like to object to introducing TS because of the following reasons:
- Why do we want type checking and accept boilerplates from the beginning, while we can easily track types and flows happening in this small project
  * type definitions for things in Atom ecosystem are not free -- we need to define
them by ourselves since the ecosystem hasn't evolved with TS
- Why can't we just convert old coffees into JS ? AFAIU your "code overhaul" doesn't need to be TS

I admit I'm one of those "dynamically typing" ppl, but I would like to avoid prefering one tool over another just because "it seems popular and promising".

Actually I don't have such a strong opinion for this package as far as the change works, since this package is kinda stable.
But I would really want to consider those tradeoffs for the other packages in our ecosystem

aminya commented:
In the following, I will respond to your points (read with a sense of humor 😉 )


> Cons:
> 
>     * boilerplates: type declarations themselves are sometimes super annoying, e.g. when you develop React code within TS, at least to some people
> 
>     * incomplete (gradual) type safety: more safety and utilities we want, more boilerplates we need to accept
>     * additional development complexity

This is the price we pay for performance. Coming from a Julia/C/C++ background, I don't see this as pain but as an opportunity to improve. 

>       
>       * since Atom doesn't come with native TS transpiler, we need to build each time we modify our code base
	
Isn't this what developers have done for many years? I don't think compiling is that scary.

>       * there might be a confusion arising around the difference between acually developed code (TS) and acually loaded code (JS)
> 

Then if you think that eventually any code is compiled to machine code, you will reconsider your concern. You should accept at some level that you don't need to understand what is actually going on underneath the program.

I haven't seen any issue with TypeScript Compiler. If it wasn't reliable Microsoft didn't use it for VSCode.

>     * if we only want linter, completion etc, we can just use eslint and/or throw your JS code to TS server
> 
> 

The simplest way to do this is to rename the file to `ts` and run `npm install`. I don't wanna deal with making a server, do you? 😄 

> To me, honestly, I would like to object to introducing TS because of the following reasons:
> 
>     * Why do we want type checking and accept boilerplates from the beginning, while we can easily track types and flows happening in this small project
>       

It is all about the performance. That is why Julia runs like C. 😄 

https://android.jlelse.eu/magic-lies-here-statically-typed-vs-dynamically-typed-languages-d151c7f95e2b

>       * type definitions for things in Atom ecosystem are not free -- we need to define them by ourselves since the ecosystem hasn't evolved with TS

- First, you don't need to use any type. TypeScript is just JavaScript. [Remember my comment](https://github.com/JunoLab/atom-julia-client/pull/705#issuecomment-605601495)

- There is `@types/atom` with very good support for all of Atom's public API.

-  Regarding the external package, We should start from somewhere.  I have made types for `etch` and `atom-select-list` already. 


> 
>     * Why can't we just convert old coffees into JS ? AFAIU your "code overhaul" doesn't need to be TS
> 
> 

Then, I just renamed js to ts:
```
decaffeinate . --loose
js-to-ts-converter .
```
Is the problem with the 2nd command? It isn't that hard to copy-paste! 😉 

> I admit I'm one of those "dynamically typing" ppl, but I would like to avoid prefering one tool over another just because "it seems popular and promising".
> 
> Actually I don't have such a strong opinion for this package as far as the change works, since this package is kinda stable.
> But I would really want to consider those tradeoffs for the other packages in our ecosystem

Thanks for summarizing all of this.

aviatesk commented:
Your points don't make sense.

> Then if you think that eventually any code is compiled to machine code, you will reconsider your concern. You should accept at some level that you don't need to understand what is actually going on underneath the program.

No, we're talking on a level of Atom's package API.
Think when you debug code with chrome dev tool. We have to debug on generated JS code although what we need to modify is original TS code.

> The simplest way to do this is to rename the file to ts and run npm install. I don't wanna deal with making a server, do you? 😄
As I repeatedly said, _throw your JS to TS server_. In other word, open your JS in VSCode.

> It is all about the performance. That is why Julia runs like C. 😄
In general, performance optimazation should be done only when there _is_ a performance issue.
Micro optimization is fun for who wrote tke code but it just smells to others who read it.
Does this package have performance issue at all ?Does atom-select-list really help loading time (in a cost of adding additional etch dependency) ?
How does TS become more performant (they compiled into JS anyway) ? Using TS
because of minor `for` loop conversion is like....

And as a side note, Julia is fast because it can infer runtime types (in a inter procedural way). This is very different approach to achieve speed from other statically typed langs like C (whenever they have intraprocedural type inference like TS).

> First, you don't need to use any type. TypeScript is just JavaScript. Remember my comment

Then why TS...
Whole the point of TS (and gradual typing in general) is to improve code base and dev experience by providing additional type information...

pfitzseb commented:
If we go with TS for this package (which I'm not at all opposed to), then *only TS files* should be in the repo. All transpiled JS should be `.gitignore`d. Conversely, *only the transpiled JS* should be `apm publish`ed (by properly setting `files` in `package.json`).

I very much doubt that TS can get us any performance improvements that we couldn't get by transpiling the current JS code ahead of time (instead of letting Atom handle that). Either way we add a build step.

Also, I'm actually seeing an *increase* in package activation time (from ~97ms to ~120ms) with this PR.

> Decaffeination is done using a single command.

No, not really -- no one wants to read autogenerated decaffeinated code (at least for any moderately complex project, which indent-detective admittedly isn't).

aminya commented:
> 
> 
> If we go with TS for this package (which I'm not at all opposed to), then _only TS files_ should be in the repo. All transpiled JS should be `.gitignore`d. Conversely, _only the transpiled JS_ should be `apm publish`ed (by properly setting `files` in `package.json`).
> 

Does this work in apm? I have not done it before:
```
"publish": "apm run build"
```

> I very much doubt that TS can get us any performance improvements that we couldn't get by transpiling the current JS code ahead of time (instead of letting Atom handle that). Either way we add a build step.
> 
> Also, I'm actually seeing an _increase_ in package activation time (from ~97ms to ~120ms) with this PR.
> 

Which version of Atom are you on? Please don't use 1.47 Nightly 3. 
https://github.com/atom/atom/issues/20531

I have already done serious benchmarks. The code is still there in the code. I think there is something wrong with your Atom config.

> > Decaffeination is done using a single command.
> 
> No, not really -- no one wants to read autogenerated decaffeinated code (at least for any moderately complex project, which indent-detective admittedly isn't).

(decaffeinate)[https://github.com/decaffeinate/decaffeinate] generates a very good and readable code. That's what I used in this package and also Julia-Client.

It doesn't use the CoffeeScript compiler itself (which isn't readable) 

Try it live: https://decaffeinate-project.org/repl/

pfitzseb commented:
> Which version of Atom are you on?

Atom 1.45. I'm getting
```
indent detective activation  115.90000000433065  ms
```
from your benchmark code.

> decaffeinate generates a very good and readable code. 

We'll have to disagree there. Of course that isn't decaffeinate's fault -- CoffeeScript just promotes coding patterns that look awful in plain JS and should be fixed during the transition.

aminya commented:
> 
> 
> > Which version of Atom are you on?
> 
> Atom 1.45. I'm getting
> 
> ```
> indent detective activation  115.90000000433065  ms
> ```
> 
> from your benchmark code.
> 

On Atom Beta:
Before (51ms):
![2020-03-30 03_45_38-Timecop — C__Users_yahyaaba_Documents_GitHub_atom-indent-detective — Atom Beta](https://user-images.githubusercontent.com/16418197/77893046-0e7d0c80-7239-11ea-90e5-dabfa90cc25b.jpg)

After (20ms):
![2020-03-30 03_44_24-Timecop — C__Users_yahyaaba_Documents_GitHub_atom-indent-detective — Atom Beta](https://user-images.githubusercontent.com/16418197/77893039-0c1ab280-7239-11ea-925d-b3358d015c1b.jpg)

aminya commented:
OK, I think I found a way inspired by https://github.com/idleberg/generator-atom-package-typescript (no Webpack option)
```
    "prepublishOnly": "npm run build",
```

aviatesk commented:
> If we go with TS for this package (which I'm not at all opposed to), then _only TS files_ should be in the repo. All transpiled JS should be `.gitignore`d. Conversely, _only the transpiled JS_ should be `apm publish`ed (by properly setting `files` in `package.json`).

Why do we want to go with TS ?
If we just want loading performance, it would be enough to translate coffees and babel JS into (literately) plain JS [as the Atom's core packages do](https://github.com/atom/atom/tree/master/packages), no ?: 
I don't see the reasons to add another layer of complexity to our development process for this package (like configurations of the build process and adding type definitions, etc)

pfitzseb commented:
Why? Dunno, I'd take it as an opportunity to learn a bit of typescript ¯\\\_(ツ)\_/¯ Imho there is no good reason from a technical point of view unless we also actively improve the code/make it more idiomatic TS.

indent-detective is basically "done" as a package, so running `npm build` occasionally doesn't seem very annoying.

aviatesk commented:
yeah, I neither have a strong opinion on TS migration for this package since it's "done" as you said.
I should have clarified that I wanted to be a bit nervous on this just because I'm wondering whether we want to migrate to TS even in other packages in our ecosystem.

I _feel_ I would like to reject to migration to TS in julia-client and ink (at least in the near future), but still not sure about the tradeoff between the cost and the benefit of migration. Actually one month before I wondered this _might_ be a good change to move our code base to TS.

aviatesk commented:
FWIW, let me talk about my personal experience. I've started to feel that I'm talking about a religion :p 

So far I wrote two Atom packages in TS (even though one is not even published)
- https://github.com/aviatesk/atom-open-in-app
- https://github.com/aviatesk/atom-hydrogen-studio

And I found introducing TS into a small package was "too much" and just verbose (in [open-in-app](https://github.com/aviatesk/atom-open-in-app)), and I finally converted to JS.
For bigger package like [Hydrogen-Studio](https://github.com/aviatesk/atom-hydrogen-studio) (and julia-client and ink), I felt TS was certainly useful in some parts, but atst I felt type declarations are really verbose when code changes drastically.
I also found that I only wanted fancy completions etc. and they're actually available for JSs, and so now I _feel_ JS is preferable, at least for me.

aminya commented:
So:
- I removed the generated files. 
- Renamed lib_src to src
- Added prepublishOnly command
- Enabled source map for debugging

aminya commented:
> 
> 
> FWIW, let me talk about my personal experience. I've started to feel that I'm talking about a religion :p
> 
> So far I wrote two Atom packages in TS (even though one is not even published)
> 
>     * [aviatesk/atom-open-in-app](https://github.com/aviatesk/atom-open-in-app)
> 
>     * [aviatesk/atom-hydrogen-studio](https://github.com/aviatesk/atom-hydrogen-studio)
> 
> 
> And I found introducing TS into a small package was "too much" and just verbose (in [open-in-app](https://github.com/aviatesk/atom-open-in-app)), and I finally converted to JS.
> For bigger package like [Hydrogen-Studio](https://github.com/aviatesk/atom-hydrogen-studio) (and julia-client and ink), I felt TS was certainly useful in some parts, but atst I felt type declarations are really verbose when code changes drastically.
> I also found that I only wanted fancy completions etc. and they're actually available for JSs, and so now I _feel_ JS is preferable, at least for me.

I checked your repositories. 
- `atom-open-in-app` was very simple, and I think you only needed 1 or 2 type definition for function inputs!
- `atom-hydrogen-studio`: I wouldn't use `import * as Atom from "atom";`. You can use `import {TextEditor} from "atom"` for example, which makes the code simple, more modular, and faster when treeshaked. The IDE extensions automatically add to imports when you use something from Atom.

aminya commented:
So:
- I removed the generated files. 
- Renamed lib_src to src
- Added `prepare` command for building before publishing and on local `npm install`.
- Enabled source map for debugging (thanks to https://stackoverflow.com/a/37799745/7910299)
- Excluding ts files from the published package (by using `files` in package.json`.

I followed the following link:
https://itnext.io/step-by-step-building-and-publishing-an-npm-typescript-package-44fe7164964c#2027

aminya commented:
I did some research, and unfortunately, `apm` doesn't support building before publishing. So there are two methods:

1) including built files in the repository
2) using Webpack as mentioned in https://github.com/atom/apm/issues/498#issuecomment-602820864. This method is the modern bundling method, [Microsoft also recommends it for VSCode extensions](https://code.visualstudio.com/api/working-with-extensions/bundling-extension).

So I will set up the 2nd option if no one disagrees.

pfitzseb commented:
Yeah, I just saw that thread on Slack. Bummer.

Webpack does seem like a good option (and might actually get us better performance), so it would be awesome if you could take a stab at implementing that.

aminya commented:
Hooray! 🚀  Just look at this. The loading time is now **6ms** using Webpack:

![image](https://user-images.githubusercontent.com/16418197/78126937-e75d4100-73d8-11ea-99c1-a3aae46eb9c3.png)

But still I don't know if this solves `apm` issue 😄

aviatesk commented:
awesome, can you get that loading time consistantly ?

aminya commented:
> 
> 
> awesome, can you get that loading time consistantly ?

Yes! These are some of them:
![image](https://user-images.githubusercontent.com/16418197/78128470-9dc22580-73db-11ea-9198-83199dcc1939.png)
![image](https://user-images.githubusercontent.com/16418197/78128518-b03c5f00-73db-11ea-97ec-bbde346a2e73.png)
![image](https://user-images.githubusercontent.com/16418197/78128493-a87cba80-73db-11ea-98db-7ca3d5549fe5.png)
![image](https://user-images.githubusercontent.com/16418197/78128540-b8949a00-73db-11ea-8a93-fddd0ebcab6c.png)
![image](https://user-images.githubusercontent.com/16418197/78128591-ccd89700-73db-11ea-9881-6ba6015f880d.png)

There is noise, but this is regular with all the packages.

pfitzseb commented:
Ok, I'm not getting any improvements to package load time, again, on both my Windows and my Linux machine. At this point I'm not sure how much I should trust the numbers I'm getting though:
![grafik](https://user-images.githubusercontent.com/6735977/78162883-d697e400-7447-11ea-806c-5182705902b9.png)

@aviatesk Can you try benchmarking this?

aminya commented:
> 
> 
> Ok, I'm not getting any improvements to package load time, again, on both my Windows and my Linux machine. At this point I'm not sure how much I should trust the numbers I'm getting though:
> ![grafik](https://user-images.githubusercontent.com/6735977/78162883-d697e400-7447-11ea-806c-5182705902b9.png)
> 
> @aviatesk Can you try benchmarking this?

This is strange. What packages are you using?

aminya commented:
So I wrote a bash script to fix this issue! No Github actions is needed.
https://github.com/aminya/atom-indent-detective/blob/TypeScript/ppublish.sh

Add  "ppublish": "bash ./ppublish.sh" to scripts and
Jut run npm run ppublish before doing apm publish

aviatesk commented:
I did bunch of reloads switching between this branch and `master` and here is the benchmark:
- on `master`: 70~80ms
- on `TypeScript`: 100~110ms

And I would also like to point that your branch has at least one regression where the previously focused editor doesn't recover its focus after indent-selector modal destroyed.

***

suggestion: let's focus on one thing; since this PR has lots of changes, we can't detect the source of the problem easily. Why not split thing into each PR to do:
- fix the bug
- change atom-space-pen-view to atom-select-list (which I don't think is necessary though)
- introduce webpack
- maybe introduce TS (if you want)

aminya commented:
Since I couldn't reproduce the result you are reporting I recorded a video, doing a live benchmark. Click on it to make it bigger:

![indent-benchmark](https://user-images.githubusercontent.com/16418197/78298908-e02d5480-74f8-11ea-8777-cefeed89874f.gif)


I did the current benchmark on Dell Latitude 5590 (with an SK hynx SC311 SATA 256GB)

I will test this on another system just in case.

aminya commented:
>     * fix the bug
> 

PR got created.

>     * change atom-space-pen-view to atom-select-list (which I don't think is necessary though)
> 
>     * maybe introduce TS (if you want)

These two are tied together. I don't wanna rewrite the whole package in JS again. 

> 
>     * introduce webpack

I can do this. It started from https://github.com/JunoLab/atom-indent-detective/pull/17/commits/6b5eadfdb7a33c0e694160b4640e0ef735213f25

aminya commented:
I added tests and benchmarks ⏲ ⌚️

aminya commented:
Could you the benchmark again? I cannot get reproduce your result but I just wanna see if it is improved on your computers.
@pfitzseb  @aviatesk

pfitzseb commented:
I'm getting 0ms on both of my machines as well now -- is that just because package activation is deferred because of the activation hook?

aminya commented:
> 
> 
> I'm getting 0ms on both of my machines as well now -- is that just because package activation is deferred because of the activation hook?

Yes. That is why. But before https://github.com/JunoLab/atom-indent-detective/pull/17/commits/525aee15704119cb5b4f3d245caacda13a63dcf5 still I got a very good loading time. 

My Github action shows `20ms` in the online benchmark: https://github.com/aminya/atom-indent-detective/runs/561525274?check_suite_focus=true#step:6:83

aminya commented:
@pfitzseb  Do you need anything else here from me?
